### PR TITLE
Implement Newton's method for iterative scheme for Poisson solve in MFIS

### LINKED ISSUE #9: Iterative scheme for MFIS simulation does not converge when applied voltage is high
For problems containing ferroelectric, dielectric, and semiconductor layers we compute self-consistent rho and phi using an iterative scheme to achieve true first or second order accuracy. Without this scheme rho computation is done using phi at the previous time step which reduces the overall order of accuracy. 

When the Dirichlet BC for Poisson solve is set to Phi_BC_lo = 0.0 and Phi_BC_hi = 1.2 to get an applied voltage of 1.2V across the MFIS stack, the iterative scheme does not converge. The error oscillates between two values rather than going down below a user specified tolerance which is set to be 1.0e-5.

The issue can be reproduced using the following script.

```
n_cell = 64 64 32

max_grid_size = 32

prob_type = 2

P_BC_flag_lo = 0 # Polarization BC at FE-DE interface
P_BC_flag_hi = 1 # Polarization BC at FE-Metal interface

Phi_Bc_lo = 0.0  # Dirichlet BC for Poisson solve at bottom metal contact
Phi_Bc_hi = 1.2   # Dirichlet BC for Poisson solve at top metal contact

TimeIntegratorOrder = 1

nsteps = 1
plot_int = 1

dt = 2.0e-13

prob_lo = -16.e-9 -16.e-9 0.e-9
prob_hi =  16.e-9  16.e-9 16.e-9

epsilon_0 = 8.85e-12
epsilonX_fe = 24.0
epsilonZ_fe = 24.0
epsilon_de = 3.9
epsilon_si = 11.7
alpha = -2.5e9
beta = 6.0e10
gamma = 1.5e11
BigGamma = 100
g11 = 1.0e-9
g44 = 1.0e-9

SC_lo = 0.0
SC_hi = 10.0e-9
DE_lo = 10.0e-9
DE_hi = 11.0e-9
FE_lo = 11.0e-9
FE_hi = 16.0e-9

lambda = 3.0e-9

```

### PR DESCRIPTION
**Pull request by prkkumar**
Motivated by issue https://github.com/AMReX-Microelectronics/FerroX/issues/9. This PR implements a new iterative algorithm for the solution of Poisson's equation for MFIS stack. It's based on the Newton's method described in the attached pdf and is more robust than our current "iterate in place" approach. 
[Newton_accelerated_Poisson_Solve_FerroX_v1.pdf](https://github.com/AMReX-Microelectronics/FerroX/files/9586472/Newton_accelerated_Poisson_Solve_FerroX_v1.pdf)

Input file to test
```
n_cell = 64 64 32

max_grid_size = 32

prob_type = 2 

P_BC_flag_lo = 0 # Polarization BC at FE-DE interface
P_BC_flag_hi = 1 # Polarization BC at FE-Metal interface

Phi_Bc_lo = 0.0  # Dirichlet BC for Poisson solve at bottom metal contact
Phi_Bc_hi = 1.2   # Dirichlet BC for Poisson solve at top metal contact

TimeIntegratorOrder = 1 

nsteps = 1500
plot_int = 500 

dt = 2.0e-13

delta = 1.e-6 #To calculate Jacobian in Newton's method for Poisson solve

prob_lo = -16.e-9 -16.e-9 0.e-9
prob_hi =  16.e-9  16.e-9 16.e-9

epsilon_0 = 8.85e-12
epsilonX_fe = 24.0
epsilonZ_fe = 24.0
epsilon_de = 3.9 
epsilon_si = 11.7
alpha = -2.5e9
beta = 6.0e10
gamma = 1.5e11
BigGamma = 100 
g11 = 1.0e-9
g44 = 1.0e-9

SC_lo = 0.0 
SC_hi = 10.0e-9
DE_lo = 10.0e-9
DE_hi = 11.0e-9
FE_lo = 11.0e-9
FE_hi = 16.0e-9

lambda = 3.0e-9
```

### CODE REVIEW

File: Source/main.cpp [hunk 1]
```diff
@@ -63,6 +63,9 @@ void main_main ()
     Real DE_lo, DE_hi, FE_lo, FE_hi, SC_lo, SC_hi;
     Real lambda;
 
+    //delta for calculating Jacobian in Newton's method for iterative Poisson solve in MFIS
+    Real delta;
```
Comment (RevathiJambunathan) [inline Source/main.cpp]:
```suggestion
    Real delta = 1.e-6;
```

File: Source/main.cpp [hunk 2]
```diff
@@ -133,6 +136,9 @@ void main_main ()
         // time step
         pp.get("dt",dt);
 
+        delta = 1.e-6;
```
Comment (RevathiJambunathan) [inline Source/main.cpp]:
```suggestion
```

File: Source/main.cpp [hunk 3]
```diff
@@ -264,7 +275,7 @@ void main_main ()
                  beta_face[2].define(convert(ba,IntVect(AMREX_D_DECL(0,0,1))), dm, 1, 0););
     
     // set cell-centered alpha coefficient to zero
-    alpha_cc.setVal(0.);
+    //alpha_cc.setVal(0.);
```
Comment (ajnonaka) [inline Source/main.cpp]: The MultiFabs "alpha_cc" and "f_prime" are redundant.  Use one or the other, and delete instances of the one you aren't using.

For what it's worth, the permittivity is stored in "beta_face" so to be consistent, I'd get rid of f_prime and just use alpha_cc.

File: Source/main.cpp [hunk 4]
```diff
@@ -281,13 +292,12 @@ void main_main ()
     mlabec.setLevelBC(0, &PoissonPhi);
     
     // (A*alpha_cc - B * div beta grad) phi = rhs
-    mlabec.setScalars(0.0, 1.0); // A = 0.0, B = 1.0
-    mlabec.setACoeffs(0, alpha_cc); //First argument 0 is lev
+    mlabec.setScalars(-1.0, 1.0); // A = 1.0, B = 1.0
```
Comment (ajnonaka) [inline Source/main.cpp]:
```suggestion
    mlabec.setScalars(-1.0, 1.0); // A = -1.0, B = 1.0; solving (-A*alpha - div beta grad) phi = RHS
```

File: Source/main.cpp [hunk 5]
```diff
@@ -281,13 +292,12 @@ void main_main ()
...
     mlabec.setBCoeffs(0, amrex::GetArrOfConstPtrs(beta_face));
 
     //Declare MLMG object
     MLMG mlmg(mlabec);
-//    mlmg.setVerbose(2);
+    //mlmg.setVerbose(2);
```
Comment (ajnonaka) [inline Source/main.cpp]: Make mlmg_verbosity a parameter you can read in from the inputs file and pass it in here - useful for debugging.

File: Source/main.cpp [hunk 6]
```diff
@@ -606,6 +666,8 @@ void main_main ()
             WriteSingleLevelPlotfile(pltfile, Plt, {"P","Phi","PoissonRHS","Ex","Ey","Ez","holes","electrons","charge"}, geom, time, step);
         }
 
+    }
+
         // MultiFab memory usage
```
Comment (ajnonaka) [inline Source/main.cpp]: Fix indentation starting from here to the bottom.

File: Source/main.cpp [hunk 7]
```diff
@@ -227,6 +233,11 @@ void main_main ()
     MultiFab e_den(ba, dm, 1, 0);
     MultiFab charge_den(ba, dm, 1, 0);
 
+    //Related to Newtons method
+    MultiFab PoissonPhi_plus_delta(ba, dm, 1, 0);
+    MultiFab PoissonRHS_phi_plus_delta(ba, dm, 1, 0);
```
Comment (ajnonaka) [inline Source/main.cpp]: You could make these local MultiFabs within dF_dPhi (simplify main, and pass around less stuff)

File: Source/ElectrostaticSolver.H [hunk 8]
```diff
@@ -48,3 +48,38 @@ void SetPhiBC_z(MultiFab& PossonPhi,
                 amrex::GpuArray<int, 3> n_cell,
                 Real                    Phi_Bc_lo,
                 Real                    Phi_Bc_hi);
+
+void dF_dPhi(MultiFab&      f_prime,
+             MultiFab&      PoissonRHS, 
+             MultiFab&      PoissonRHS_phi_plus_delta,
```
Comment (ajnonaka) [inline Source/ElectrostaticSolver.H]: General style comment - with so many arguments it's probably useful to use "const MultiFab&" in instances where the input parameter is meant to be read-only.  This change would have to propagate across many functions in the code (you can't just fix it here, you have to fix it within all the functions called within as well).  It's kind of hard to tell exactly what this function is doing without either the "consts" or a comment at the top explicitly stating what this function is doing (e.g., "Compute f_prime = ..."

**Review by ajnonaka — CHANGES_REQUESTED** [review]
Looks pretty good - some of my comments should be addressed, some can wait for a future PR (you'll know which are which).

File: Source/main.cpp (hunk 7 above)
Comment (prkkumar) [reply Source/main.cpp]: done.

File: Source/main.cpp (hunk 6 above)
Comment (prkkumar) [reply Source/main.cpp]: done.

File: Source/main.cpp (hunk 5 above)
Comment (prkkumar) [reply Source/main.cpp]: done. I set the default to 1 for cleaner stdout. We can set it to 2 through input file for debugging etc..

File: Source/main.cpp (hunk 3 above)
Comment (prkkumar) [reply Source/main.cpp]: Done. Going with alpha_cc as you suggested.

File: Source/ElectrostaticSolver.H (hunk 8 above)
Comment (prkkumar) [reply Source/ElectrostaticSolver.H]: Should we address it in the next PR? I also plan to use namespace for storing constants that I pass around as input parameters in many functions. 

Next PR could be devoted to clean up issues to address this comment and to use namespace (https://github.com/AMReX-Microelectronics/FerroX/issues/18).

File: Source/ElectrostaticSolver.cpp [hunk 9]
```diff
@@ -79,6 +80,92 @@ void ComputePoissonRHS(MultiFab&               PoissonRHS,
...
+            const Array4<Real>& alpha = alpha_cc.array(mfi);
+            const Array4<Real>& phi = PoissonPhi.array(mfi);
+            const Array4<Real>& poissonRHS = PoissonRHS.array(mfi);
+            const Array4<Real>& poissonRHS_phi_plus_delta = PoissonRHS_phi_plus_delta.array(mfi);
+
+            amrex::ParallelFor( bx, [=] AMREX_GPU_DEVICE (int i, int j, int k)
+            {
+                   alpha(i,j,k) = 1./delta*(poissonRHS_phi_plus_delta(i,j,k) - poissonRHS(i,j,k));
+            });
+        }
```
Comment (ajnonaka) [inline Source/ElectrostaticSolver.cpp:147]:
```suggestion
        LinComb(alpha_cc, 1./delta, PoissonRHS_phi_plus_delta, 0, -1./delta, PoissonRHS, 0, 0, 1, 0);
```

Comment (ajnonaka) [inline Source/ElectrostaticSolver.cpp]:
```suggestion
        LinComb(alpha_cc, 1./delta, PoissonRHS_phi_plus_delta, 0, -1./delta, PoissonRHS, 0, 0, 1, 0);
```

File: Source/main.cpp [hunk 10]
```diff
@@ -264,7 +271,7 @@ void main_main ()
                  beta_face[2].define(convert(ba,IntVect(AMREX_D_DECL(0,0,1))), dm, 1, 0););
     
     // set cell-centered alpha coefficient to zero
-    alpha_cc.setVal(0.);
+    //alpha_cc.setVal(0.);
```
Comment (ajnonaka) [inline Source/main.cpp]: Delete lines 272-274

**Review by ajnonaka — APPROVED** [review]
Approve pending the two suggestions.

**Pull request merged**